File: leetcode692/leetcode692_my_version.cpp
class Solution {
public:
    vector<string> topKFrequent(vector<string>& words, int k) {
        unordered_map<string, int> freq;
        
        for (auto word : words) {
            ++freq[word];
        }
        
        vector<pair<int, string>> ans(freq.size());
        int i = 0;
        for (auto word_fre : freq) {
            ans[i] = {word_fre.second, word_fre.first};
            ++i;
        }
        
        sort(begin(ans), end(ans), comp);
        
        vector<string> output(k);
        
        for (i = 0; i < k; ++i) {
            output[i] = ans[i].second;
        }
        
        return output;
    }
    
private:
    static bool comp(pair<int, string>& p1, pair<int, string>& p2) {
        return p1.first > p2.first || (p1.first == p2.first && p1.second < p2.second);
    }
};

// Runtime: 12 ms, faster than 97.17% of C++ online submissions for Top K Frequent Words.
// Memory Usage: 11.2 MB, less than 100.00% of C++ online submissions for Top K Frequent Words.
// time complexity: O(n log n)
// space complexity: O(n)

File: leetcode692/leetcode692_my_version.java
class Solution {
    public List<String> topKFrequent(String[] words, int k) {
        Map<String, String> freq = new HashMap<>();

        for (String w : words) {
            if (freq.get(w) == null) {
                freq.put(w, "1");
            } else {
                int num = Integer.valueOf(freq.get(w)) + 1;
                freq.put(w, String.valueOf(num));
            }
        }

        PriorityQueue<String[]> pq = new PriorityQueue<>(
                (a, b) -> (Integer.valueOf(a[1]) == Integer.valueOf(b[1]) ? b[0].compareTo(a[0])
                        : Integer.valueOf(a[1]) - Integer.valueOf(b[1])));

        for (String key : freq.keySet()) {
            pq.add(new String[] { key, freq.get(key) });

            if (pq.size() > k) {
                pq.poll();
            }
        }

        List<String> ans = new ArrayList<>();
        for (int i = 0; i < k; i++) {
            ans.add(pq.poll()[0]);
        }
        // reverse
        for (int i = 0; i < ans.size() / 2; i++) {
            String tmp = ans.get(i);
            ans.set(i, ans.get(ans.size() - 1 - i));
            ans.set(ans.size() - 1 - i, tmp);
        }

        return ans;
    }
}

// Runtime: 7 ms, faster than 27.29% of Java online submissions for Top K
// Frequent Words.
// Memory Usage: 39 MB, less than 69.10% of Java online submissions for Top K
// Frequent Words.
// time complexity: O(n log n)
// space complexity: O(n)